staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -158,28 +291,94 @@ func (q *delayingType) ShutDown() {
 	})
 }
 
-// AddAfter adds the given item to the work queue after the given delay
+// Add follows the same behaviour as Add() on Interface(), except that it also
+// removes a duplicate item from the 'waiting' queue, if it exists.
+// All calls to Add are synchronous to preserve backwards compatibility with existing clients.
+func (q *delayingType) Add(item interface{}) {
+	q.sendItemToWaitingLoop(item, DelayingOptions{Duration: 0}, waitingLoopFeatures{synchronous: true})
+}
+
+// AddAfter adds the given item to the work queue after the given delay, or dropped if the same item is already
+// queued in the 'active' queue.  Where the item is already in the 'waiting' queue then it is updated
+// if the new delay is shorter.
+// Calls to AddAfter() are synchronous when delay <= 0 to preserve compatibility with existing clients.
 func (q *delayingType) AddAfter(item interface{}, duration time.Duration) {
+	q.sendItemToWaitingLoop(item, DelayingOptions{Duration: duration}, waitingLoopFeatures{synchronous: (duration <= 0)})
+}
+
+// AddWithOptions is a new method that add's items to the delayed queue.
+// Calls are asynchronous so that the client can return quicker and continue with useful work.
+// The DelayingOptions give the client more control over the queueing behaviour.
+func (q *delayingType) AddWithOptions(item interface{}, opts DelayingOptions) {
```

Comment:
I wouldn't expect the two to work the same way for these reasons: -
* The new WithOptions calls makes synchronous/asynchronous an explicit option on the call and it would be confusing to mix explicit with implicit behaviour.
* Add/AddAfter can be viewed as the old interface, maintained for backwards compatibility, but AddWithOptions is intended to replace them.  Keeping the backwards compatibility allows us to keep the old behaviours that they provide, so that we don't need to refactor everything in one go but having the new interface (AddWithOptions) gives us a chance to expose the new different richer behaviour.
* With the new implementation there is now a performance penalty to making a synchronous add because it has to be processed though the waitLoop where it didn't before.  Unless you really need synchronous behaviour (in which case you can just ask for it in the options) then the default should be the more performant asynchronous add which returns back sooner and lets the client carry on with their work.

Reply:
Oh that's what comes from re-writing it 50 times! :)  I took the Synchronous option out again because you only really want this option in testing when you then want to go on and assert that something has happened.  I removed it because the type `TestDelayingInterface` turns it on by default for anyone wanting to write tests directly against the delaying queue.

I'm completely happy for there to be more discussion around this, we could make it a set-able feature of the delaying queue and put it back into the DelayingOptions (where I did actually have it for a while), or we can take the view that nobody apart from a test would want to make synchronous calls and maybe its better to state that all calls are asynchronous unless using the test version.

Reply:
I see :). Hard to envision if there is some scenario where synchronous behaviour might be useful. We probably should not let other depend on the synchronicity in their own tests.  

Let's not add the `synchronous` field to the API unless we really need it to make the API simpler. If there is a use case we can always add it later, since `AddWithOptions` has no users yet.